src/stream/src/executor/merge.rs

```diff
@@ -385,29 +353,37 @@ impl Execute for MergeExecutor {
     }
 }
 
+type BoxedMessageInput<InputId, M> = BoxedInput<InputId, MessageStreamItemInner<M>>;
+
 /// A stream for merging messages from multiple upstreams.
-pub struct SelectReceivers {
+/// Can dynamically add and delete upstream streams.
+/// For the meaning of the generic parameter `M` used, refer to `BarrierInner<M>`.
+pub struct DynamicReceivers<InputId, M> {
     /// The barrier we're aligning to. If this is `None`, then `blocked_upstreams` is empty.
-    barrier: Option<DispatcherBarrier>,
+    barrier: Option<BarrierInner<M>>,
     /// The upstreams that're blocked by the `barrier`.
-    blocked: Vec<BoxedInput>,
+    blocked: Vec<BoxedMessageInput<InputId, M>>,
```

Comment:
Can try making `upstream_input_ids` as `fn upstream_input_ids(&self) -> impl Iterator<Item = InputId> + '_`? It can be implemented as `blocked.iter().map(...).chain(active.iter().map(...))`, and then we don't have to maintain a separate `upstream_input_ids`.